applications/porthole/src/lib.rs

```diff
@@ -0,0 +1,1042 @@
+            x: self.x + other.x,
+            y: self.y + other.y,
+        }
+    }
+}
+
+impl Add<Rect> for ScreenPos {
+    type Output = Self;
+
+    fn add(self, other: Rect) -> Self {
+        Self {
+            x: self.x + other.x as i32,
+            y: self.y + other.y as i32,
+        }
+    }
+}
+
+#[derive(Clone, Copy, Debug)]
+pub struct Rect {
+    pub width: usize,
+    pub height: usize,
+    pub x: isize,
+    pub y: isize,
+}
+
+impl Rect {
+    fn new(width: usize, height: usize, x: isize, y: isize) -> Rect {
+        Rect {
+            width,
+            height,
+            x,
+            y,
+        }
+    }
+
+    fn x_plus_width(&self) -> isize {
+        self.x + self.width as isize
+    }
+
+    fn y_plus_height(&self) -> isize {
+        self.y + self.height as isize
+    }
+
+    fn detect_collision(&self, other: &Rect) -> bool {
+        self.x < other.x_plus_width()
+            && self.x_plus_width() > other.x
+            && self.y < other.y_plus_height()
+            && self.y_plus_height() > other.y
+    }
+
+    /// Creates a new `Rect` from visible parts of itself.
+    pub fn visible_rect(&self) -> Rect {
+        let mut x = self.x;
+        let y = self.y;
+        let mut width = self.width as isize;
+        let mut height = self.height as isize;
+        if self.x < 0 {
+            x = 0;
+            width = self.x_plus_width();
+        } else if self.x_plus_width() > SCREEN_WIDTH as isize {
+            x = self.x;
+            let gap = (self.x + self.width as isize) - SCREEN_WIDTH as isize;
+            width = self.width as isize - gap;
+        }
+        if self.y_plus_height() > SCREEN_HEIGHT as isize {
+            let gap = (self.y + self.height as isize) - SCREEN_HEIGHT as isize;
+            height = self.height as isize - gap;
+        }
+        let visible_rect = Rect::new(width as usize, height as usize, x, y);
+        visible_rect
+    }
+}
+
+pub struct VirtualFrameBuffer {
+    width: usize,
+    height: usize,
+    buffer: BorrowedSliceMappedPages<u32, Mutable>,
+}
+
+impl VirtualFrameBuffer {
+    pub fn new(width: usize, height: usize) -> Result<VirtualFrameBuffer, &'static str> {
+        let kernel_mmi_ref =
+            memory::get_kernel_mmi_ref().ok_or("KERNEL_MMI was not yet initialized!")?;
+        let size = width * height * core::mem::size_of::<u32>();
+        let pages = memory::allocate_pages_by_bytes(size)
+            .ok_or("could not allocate pages for a new framebuffer")?;
+        let mapped_buffer = kernel_mmi_ref
+            .lock()
+            .page_table
+            .map_allocated_pages(pages, PteFlags::new().valid(true).writable(true))?;
+        Ok(VirtualFrameBuffer {
+            width,
+            height,
+            buffer: mapped_buffer
+                .into_borrowed_slice_mut(0, width * height)
+                .map_err(|(_mp, s)| s)?,
+        })
+    }
+
+    fn copy_windows_into_main_vbuffer(&mut self, window: &mut MutexGuard<Window>) {
```

Comment:
this should really be a Window method
```suggestion
    fn copy_into_vbuffer(&mut self, vbuf: &mut VirtualFrameBuffer){
```